File: backend/src/main/java/dev/arc/model/Definition.java
package dev.arc.model;

import java.util.List;
import java.util.Map;

/** Portable executable graph. Presentation coordinates never affect execution. */
public record Definition(
    int schemaVersion, List<Input> inputs, List<Node> nodes, List<Edge> edges, List<String> notes) {
  public Definition(int schemaVersion, List<Input> inputs, List<Node> nodes, List<Edge> edges) {
    this(schemaVersion, inputs, nodes, edges, List.of());
  }

  public record SourceBinding(
      String id, int version, Map<String, String> bindings, String pointer, String onError) {}

  public record Input(
      String name, String type, boolean required, Object defaultValue, SourceBinding source) {
    public Input(String name, String type, boolean required, Object defaultValue) {
      this(name, type, required, defaultValue, null);
    }
  }

  public record Position(double x, double y) {}

  /** Case identity stays stable when its label, predicate or priority changes. */
  public record BranchCase(String id, String label, String expression) {}

  public record Field(String name, String expression) {}

  public record Node(
      String id,
      String type,
      String label,
      Position position,
      String expression,
      String output,
      String ruleId,
      Integer version,
      Map<String, String> bindings,
      List<BranchCase> cases,
      List<Field> fields) {
    public Node(
        String id,
        String type,
        String label,
        Position position,
        String expression,
        String output,
        String ruleId,
        Integer version,
        Map<String, String> bindings) {
      this(id, type, label, position, expression, output, ruleId, version, bindings, null, null);
    }

    public boolean storesResult() {
      return "FORMULA".equals(type) || "REFERENCE".equals(type) || "TRANSFORM".equals(type);
    }
  }

  public record Edge(String id, String source, String target, String sourceHandle) {}
}

File: backend-rails/lib/arc/definition.rb
module Arc
  module Definition
    module_function
    def record(value, names, defaults = {})
      return nil if value.nil?
      raise TypeError unless value.is_a?(Hash)
      names.split.to_h { |name| [name, value.fetch(name, defaults[name])] }
    end

    def list(value)
      return nil if value.nil?
      raise TypeError unless value.is_a?(Array)
      value.map { |item| yield item }
    end

    def input(value)
      p = record(value, "name type required defaultValue source", "required" => false)
      return nil unless p
      p["source"] = record(p["source"], "id version bindings pointer onError", "version" => 0)
      p
    end

    def normalize(value)
      d = record(value, "schemaVersion inputs nodes edges notes", "schemaVersion" => 0)
      return nil unless d
      d["inputs"] = list(d["inputs"]) { |p| input(p) }
      d["nodes"] = list(d["nodes"]) do |value|
        n = record(value, "id type label position expression output ruleId version bindings")
        n["position"] = record(n["position"], "x y", "x" => 0, "y" => 0) if n
        n
      end
      d["edges"] = list(d["edges"]) { |e| record(e, "id source target sourceHandle") }
      d
    end

    def source(value)
      c = record(value, "kind url parameters entries secretHeaders timeoutMs", "timeoutMs" => 0)
      c["parameters"] = list(c["parameters"]) { |p| input(p) } if c
      c
    end
  end
end
